mission_control/navigator_missions/navigator_singleton/pose_editor.py

```diff
@@ -0,0 +1,201 @@
+from __future__ import division
+import warnings
+
+import numpy as np
+from tf import transformations
+from nav_msgs.msg import Odometry
+from uf_common.msg import PoseTwistStamped, PoseTwist, MoveToGoal
+from geometry_msgs.msg import Pose, PoseStamped, Quaternion, Point, Vector3, Twist
+from sub8_ros_tools import rosmsg_to_numpy, make_header
+
+
+UP = np.array([0.0, 0.0, 1.0], np.float64)
+EAST, NORTH, WEST, SOUTH = [transformations.quaternion_about_axis(np.pi / 2 * i, UP) for i in xrange(4)]
+UNITS = {'m': 1, 'ft': 0.3048, 'yard': 0.9144, 'rad': 1, 'deg': 0.0174533}
+
+def normalized(x):
+    x = np.array(x)
+    if max(map(abs, x)) == 0:
+        warnings.warn('Normalizing zero-length vector to random unit vector')
+        x = np.random.standard_normal(x.shape)
+    x = x / max(map(abs, x))
+    x = x / np.linalg.norm(x)
+    return x
+
+
+def get_perpendicular(a, b=None):
+    a = np.array(a)
+    if max(map(abs, a)) == 0:
+        if b is not None:
+            return get_perpendicular(b)
+        return normalized(np.random.standard_normal(a.shape))
+
+    if b is None:
+        b = np.random.standard_normal(a.shape)
+    b = np.array(b)
+    x = np.cross(a, b)
+    if max(map(abs, x)) == 0:
+        return get_perpendicular(a)
+    return normalized(x)
+
+
+def triad((a1, a2), (b1, b2)):
+    # returns quaternion that rotates b1 to a1 and b2 near a2
+    # can get orientation by passing in (global, local)
+    aa = get_perpendicular(a1, a2)
+    A = np.array([normalized(a1), aa, normalized(np.cross(a1, aa))])
+    bb = get_perpendicular(b1, b2)
+    B = np.array([normalized(b1), bb, normalized(np.cross(b1, bb))])
+    rot = A.T.dot(B)
+    return transformations.quaternion_from_matrix(
+        [(a, b, c, 0) for a, b, c in rot] +
+        [(0, 0, 0, 1)])
+
+
+def look_at(forward, upish=UP):
+    # assumes standard forward-left-up body coordinate system
+    return triad((forward, upish), ([1, 0, 0], UP))
+
+
+def look_at_without_pitching(forwardish, up=UP):
```

Comment:
These are ported from the sub, I will remove since they don't make sense in this context